### PR TITLE
refactor http request and add put and delete

### PR DESCRIPTION
**Pull request by charmingduchess**
Noisy PR for a couple small changes and additions:
- extracting the error handling and the basic request format into reusable functions
- refactor existing get and post methods to use reusable functions
- create new delete and put requests to use reusable functions

In order to refactor the request methods to be less repetitive, I had to use the axios.request() method, which is not supported well by the axios mocker. I had to replace the "onGet" methods with the generic "onRequest" in the tests. I don't love that, but I think it's acceptable. 

I'm also not entirely sure how to write tests for these that aren't entirely repetitive. Also any guidance on updating the integration tests would be appreciated!

### CODE REVIEW

**Comment by charmingduchess** [comment]
@nonword here's the script
```
const { config, authenticate, deleteRequest, put, post, get } = require('./index.js')

const doTheThing = async () => {
  config('test/installation-test/creds.json')
  authenticate()
  const patron = await post('/patrons', { names: ['duchess wazzinsky'] })
  const id = patron.link.split('patrons/')[1]
  let returnedPatron = await get('/patrons/' + id)
  console.log('first get', returnedPatron)
  await put('/patrons/' + id, { names: ['princess buttercup'] })
  returnedPatron = await get('/patrons/' + id + '?fields=names')
  console.log('second get', returnedPatron)
  const deleteResponse = await deleteRequest('/patrons/' + id)
  console.log(deleteResponse)
  returnedPatron = await get('/patrons/' + id)
// a patron id is still returned, but with no other data. i'm assuming this is correct behavior, but not 100% sure.
  console.log('third get', returnedPatron)
}

doTheThing()
```

**Review by nonword — COMMENTED** [review]
Nice reorganization

Seems like the integration tests have fallen out of sync with the module's api. They're also not documented. So I would propose ticketing addressing that separately. (Possibly including your own integration testing script for creating/modifying patrons.)

As for repetition in tests: It looks like you copied the POST test, which happens to also test reauthentication. Given the structure of the app, I don't think testing reauthentication functionality for each http method is necessary, personally. So you could remove the reauthentication aspect of the PUT test. Or move [the reauthentication setup](https://github.com/NYPL-discovery/sierra-wrapper/pull/16/files#diff-af95cb6f5f65fe70216ddce70325e785438d5aaff956a0666ebf46ef6402f526R180-R184) into a `beforeEach` for use by the PUT and POST tests.

File: test/index.test.js [hunk 1]
```diff
@@ -167,27 +167,47 @@ describe('test', function () {
...
+      wrapper.config({ key: credsKey, secret: credsSecret, base: credsBase })
+      const reauthenticate = wrapper.__get__('_reauthenticate')
+      const reauthenticateSpy = sinon.spy(reauthenticate)
+      wrapper.__set__('_reauthenticate', reauthenticateSpy)
+      const data = { books: ['the', 'books'] }
+      // authenticate
+      mockAxios.onPost(`${credsBase}token`, auth)
+        .reply(200, { access_token: '12345' })
+      // post() call
+      mockAxios.onAny()
```
[the remark this answers is not in the document]
Comment (charmingduchess) [reply test/index.test.js:190]: I dont think so!

**Comment by charmingduchess** [comment]
> As for repetition in tests: It looks like you copied the POST test, which happens to also test reauthentication. Given the structure of the app, I don't think testing reauthentication functionality for each http method is necessary, personally. 

Yea I agree that it's not necessary, but there isn't really anything else being tested. The methods aren't really doing any transformations to the data. I don't want to have no tests for a method though. Maybe just testing that no errors are thrown when valid parameters are passed in?

File: index.js [hunk 2]
```diff
@@ -81,42 +81,60 @@ async function authenticate (_retryCount = 1) {
...
+    return await _handleErrors(error, deleteRequest, path)
+  }
+}
+
+async function put (path, data) {
+  try {
+    const response = await _doHttpRequest('put', path, data)
+    return response.data
+  } catch (error) {
+    return await _handleErrors(error, deleteRequest, path)
```
Comment (nonword) [inline index.js:111]: This too should pass `put`, `path`, and `data`?

File: index.js [hunk 3]
```diff
@@ -81,42 +81,60 @@ async function authenticate (_retryCount = 1) {
...
+  }
+}
+
+async function post (path, data) {
+  try {
+    const response = await _doHttpRequest('post', path, data)
+    return response.data
+  } catch (error) {
+    return await _handleErrors(error, deleteRequest, path)
```
Comment (nonword) [inline index.js:102]: Shouldn't this pass `post`, `path`, and `data`?

**Comment by nonword** [comment]
> Yea I agree that it's not necessary, but there isn't really anything else being tested.

I see. Yeah ideally a test of `wrapper.put` would assert that it calls `authenticate` and uses the token in a PUT call with the expected body to the expected path. It looks like you can set up expectations on the axios-mock-adapter spies like this:

```
const putCall = mockAxios.onAny().reply(200, 'success')

expect(putCall.history.put.length).to.equal(1)
expect(putCall.history.put[0].headers.Authorization).to.equal('Bearer ...')
expect(putCall.history.put[0].method).to.equal('put')
expect(putCall.history.put[0].url).to.equal('/path')
expect(putCall.history.put[0].url).to.deepEqual({ ... })
```

File: index.js (hunk 2 above)
Comment (charmingduchess) [reply index.js:111]: yes good catch!

**Comment by charmingduchess** [comment]
@nonword those properties are only populated for specific axios.put/get/post/delete calls. I don't see generic axios.request() calls maintained anywhere on the mockAxios instance.

**Comment by charmingduchess** [comment]
@nonword updated a bunch of the tests. I organized tests for similar behavior, ie testing auth retries for all requests, testing successful requests under the same describe blocks with its for each method. 
I extracted and commented on some repeated set up logic, but I couldn't quite follow all of it, so some arcane incantations remain. I also swapped out as much of the sinon stubbing as I could stomach in an effort to streamline how tests are working. 
It might make more sense to just look at the updated testing file instead of trying to grok the diff.
The mock axios history functionality made it possible to do a lot of simplification and also proper testing, so thanks again for that!

File: test/index.test.js [hunk 4]
```diff
@@ -156,41 +142,99 @@ describe('test', function () {
...
-      // authenticate
+  describe('reauthentication for expired access token', () => {
+    let handleAuthErrorSpy
+    const path = 'path'
+    beforeEach(() => {
+      const _handleAuthError = wrapper.__get__('_handleAuthError')
+      handleAuthErrorSpy = sinon.spy(_handleAuthError)
+      wrapper.__set__('_handleAuthError', handleAuthErrorSpy)
+      // mock successful auth call
       mockAxios.onPost(`${credsBase}token`, auth)
```
Comment (nonword) [inline test/index.test.js:153]: I don't think this `onPost` is matching anything. We don't authenticate by posting a JSON body with username and password, so I don't think `mockAxios.onPost(`${credsBase}token`, auth)` is matching the token request. If you inspect the `mockAxios.history`, I think you'll see the client is not using the mocked access token.

**Comment by nonword** [comment]
Updated [sample working test](https://github.com/NYPL-discovery/sierra-wrapper/pull/16#issuecomment-1908470856) to use `onPut` instead of `onAny`. I think axios-mock-adapter actually *does* fire `onPut` when you make a `PUT` via `axios.request`

**Comment by charmingduchess** [comment]
@nonword third time's the charm... 🤞🏻

File: test/index.test.js [hunk 5]
```diff
@@ -156,41 +142,99 @@ describe('test', function () {
...
+      await wrapper.put('newBooks', data)
+      expect(handleAuthErrorSpy.calledWith(wrapper.put, path, data))
+    })
+    it('delete', async () => {
+      await wrapper.deleteRequest('newBooks')
+      expect(handleAuthErrorSpy.calledWith(wrapper.deleteRequest, path, data))
+    })
+    it('get', async () => {
+      await wrapper.get('books')
+      expect(handleAuthErrorSpy.calledWith(wrapper.get, path, data))
```
[the remark this answers is not in the document]
Comment (nonword) [reply test/index.test.js:165]: This appears to still be an issue

**Review by nonword — APPROVED** [review]
Looks good!

**Pull request merged**